### ISSUE TITLE
[x86 isel] Bad codegen for unpck and permute operations

### ISSUE THREAD

**Issue by zhuhan0** [llvm/llvm-project#61964]
https://godbolt.org/z/j7Eff1cbj shows some reduced code from a decode/decompression routine. Defining `BAD_CODEGEN` gives you really bad codegen for usage of the `permute` mask on line 31. It replaces 2 port5 instructions with 5 port5 instructions.

Inlining permute mask:
```asm
        vpermq  ymm3, ymm2, 136                 # ymm3 = ymm2[0,2,0,2]
        vpshufb ymm3, ymm3, ymm0
        vpermq  ymm4, ymm1, 136                 # ymm4 = ymm1[0,2,0,2]
        vpshufb ymm4, ymm4, ymm0
        vpunpcklwd      ymm3, ymm4, ymm3        # ymm3 = ymm4[0],ymm3[0],ymm4[1],ymm3[1],ymm4[2],ymm3[2],ymm4[3],ymm3[3],ymm4[8],ymm3[8],ymm4[9],ymm3[9],ymm4[10],ymm3[10],ymm4[11],ymm3[11]
        vpermq  ymm2, ymm2, 221                 # ymm2 = ymm2[1,3,1,3]
        vpshufb ymm2, ymm2, ymm0
        vpermq  ymm1, ymm1, 221                 # ymm1 = ymm1[1,3,1,3]
        vpshufb ymm1, ymm1, ymm0
        vpunpcklwd      ymm1, ymm1, ymm2        # ymm1 = ymm1[0],ymm2[0],ymm1[1],ymm2[1],ymm1[2],ymm2[2],ymm1[3],ymm2[3],ymm1[8],ymm2[8],ymm1[9],ymm2[9],ymm1[10],ymm2[10],ymm1[11],ymm2[11]
```
Outlining permute mask:
```asm
        vpunpcklwd      ymm3, ymm1, ymm2        # ymm3 = ymm1[0],ymm2[0],ymm1[1],ymm2[1],ymm1[2],ymm2[2],ymm1[3],ymm2[3],ymm1[8],ymm2[8],ymm1[9],ymm2[9],ymm1[10],ymm2[10],ymm1[11],ymm2[11]
        vpunpckhwd      ymm1, ymm1, ymm2        # ymm1 = ymm1[4],ymm2[4],ymm1[5],ymm2[5],ymm1[6],ymm2[6],ymm1[7],ymm2[7],ymm1[12],ymm2[12],ymm1[13],ymm2[13],ymm1[14],ymm2[14],ymm1[15],ymm2[15]
        vpermd  ymm2, ymm0, ymm3
        vpermd  ymm1, ymm0, ymm1
```

**Comment by zhuhan0** [comment]
In the bad case, `simplifyX86vpermv` realizes that the mask is a constant and transformed the permd intrinsic into a shufflevector. For example:
```
%shuffle.i59 = shufflevector <16 x i16> %arrayidx1.val60, <16 x i16> %arrayidx5.val61, <16 x i32> <i32 4, i32 20, i32 5, i32 21, i32 6, i32 22, i32 7, i32 23, i32 12, i32 28, i32 13, i32 29, i32 14, i32 30, i32 15, i32 31>
```

At the beginning of ISel the result of unpck and permute operations looks like this:
```
t62: v16i16 = vector_shuffle<4,20,12,28,5,21,13,29,6,22,14,30,7,23,15,31> t9, t14
```
which is basically the same as the intrinsics code the user had written just in DAG node format:
```
ymm0[1] = _mm256_unpackhi_epi16(ymm1[0], ymm1[1]);
ymm0[1] = _mm256_permutevar8x32_epi32(ymm0[1], permute);
```
=>
```
vpunpckhwd      ymm1, ymm1, ymm2        # ymm1 = ymm1[4],ymm2[4],ymm1[5],ymm2[5],ymm1[6],ymm2[6],ymm1[7],ymm2[7],ymm1[12],ymm2[12],ymm1[13],ymm2[13],ymm1[14],ymm2[14],ymm1[15],ymm2[15]
vpermd  ymm1, ymm0, ymm1
```
However during X86 ISel the vector_shuffle is transformed into this:
```
              t171: v4i64 = X86ISD::VPERMI t156, TargetConstant:i8<-120>
            t158: v8i32 = bitcast t171
          t146: v16i16 = bitcast t158
        t142: v32i8 = bitcast t146
      t143: v32i8 = X86ISD::PSHUFB t142, t164
    t144: v16i16 = bitcast t143
              t172: v4i64 = X86ISD::VPERMI t152, TargetConstant:i8<-120>
            t155: v8i32 = bitcast t172
          t151: v16i16 = bitcast t155
        t147: v32i8 = bitcast t151
      t148: v32i8 = X86ISD::PSHUFB t147, t164
    t149: v16i16 = bitcast t148
  t101: v16i16 = X86ISD::UNPCKL t144, t149
```
There is an existing function `lowerShuffleAsUNPCKAndPermute` in `X86ISelLowering` which should be able to hit this case. However it assumes all the mask indices are in-lane. As an example, the third index 12, being in lane one of the mask/output, does not fall into lane one of neither input (0..8, 16..24). So the transformation fails. Removing the lane restriction, makes this transformation hit and produces the same code as the good case:
```
vpunpcklwd      %ymm2, %ymm1, %ymm3     # ymm3 = ymm1[0],ymm2[0],ymm1[1],ymm2[1],ymm1[2],ymm2[2],ymm1[3],ymm2[3],ymm1[8],ymm2[8],ymm1[9],ymm2[9],ymm1[10],ymm2[10],ymm1[11],ymm2[11]
vpermd  %ymm3, %ymm0, %ymm3
vpunpckhwd      %ymm2, %ymm1, %ymm1     # ymm1 = ymm1[4],ymm2[4],ymm1[5],ymm2[5],ymm1[6],ymm2[6],ymm1[7],ymm2[7],ymm1[12],ymm2[12],ymm1[13],ymm2[13],ymm1[14],ymm2[14],ymm1[15],ymm2[15]
vpermd  %ymm1, %ymm0, %ymm1
```
Not sure what was the intention of the lane restriction. Probably for performance reasons? cc @RKSimon

**Comment by zhuhan0** [comment]
Submit an RFC fix: https://reviews.llvm.org/D147668

**Comment by RKSimon** [comment]
Fixed by ef38880ce03bc1f1fb3606c5a629151f3d0e975e
